### PR TITLE
First PR

### PR DESCRIPTION
**Pull request by apomelitos**

### CODE REVIEW

File: my-app/src/components/EvolutionsList.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import { FC, useCallback, useEffect, useState } from 'react';
+import { PokemonType, EvolutionChain, EvolutionsListProps } from '../types';
+import './EvolutionsList.scss';
+
+export const EvolutionList: FC<EvolutionsListProps> = ({ pokemonSpeciesURL }) => {
+  const [evolutions, setEvolutions] = useState<EvolutionChain[]>();
+
+  const getArrayFromRecursiveObject = (obj: EvolutionChain): Array<EvolutionChain> => {
```
Comment (TheForsakenSpirit) [inline my-app/src/components/EvolutionsList.tsx]:
```suggestion
  const getArrayFromRecursiveObject = (obj: EvolutionChain): EvolutionChain[] => {
```

File: my-app/src/components/EvolutionsList.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      const {
+        sprites: { front_default },
+      } = (await resp.json()) as PokemonType;
+      url = front_default;
+    } catch (err) {
+      console.error(err);
+    }
+
+    return url;
+  };
```
Comment (TheForsakenSpirit) [inline my-app/src/components/EvolutionsList.tsx]: Move out from render function

File: my-app/src/components/EvolutionsList.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  const fetchEvolutions = useCallback(async () => {
+    const speciesResponse = await fetch(pokemonSpeciesURL);
+    const {
+      evolution_chain: { url },
+    } = (await speciesResponse.json()) as { evolution_chain: { url: string } };
+
+    const evolutionResponse = await fetch(url);
+    const { chain } = (await evolutionResponse.json()) as { chain: EvolutionChain };
```
Comment (TheForsakenSpirit) [inline my-app/src/components/EvolutionsList.tsx]: Use [type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) to cast types.

File: my-app/src/components/EvolutionsList.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    const evolutionsList = getArrayFromRecursiveObject(chain);
+
+    for (let evolIdx = 0; evolIdx < evolutionsList.length; evolIdx++) {
+      const name = evolutionsList[evolIdx].species.name;
+      evolutionsList[evolIdx].sprite = await fetchPokemonSpriteByName(name);
+    }
```
Comment (TheForsakenSpirit) [inline my-app/src/components/EvolutionsList.tsx]: use `Promise.all` and map to make paralleled query.

File: my-app/src/components/EvolutionsList.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  useEffect(() => {
+    fetchEvolutions();
+  }, [fetchEvolutions]);
+
+  return (
+    <ul className='evolution-chain'>
+      {evolutions &&
+        evolutions.map((item, idx) => (
+          <li key={idx}>
```
Comment (TheForsakenSpirit) [inline my-app/src/components/EvolutionsList.tsx]: try to generate key from item.

File: my-app/src/components/Pagination.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  siblingCount: number;
+  totalCount: number;
+  pageSize: number;
+  currentPage: number;
+};
+
+export const Pagination: FC<PaginationProps> = ({ siblingCount = 1, totalCount, pageSize, currentPage }) => {
+  // const shouldShowRightDots = false;
+  // const shouldShowLeftDots = false;
+
```
Comment (TheForsakenSpirit) [inline my-app/src/components/Pagination.tsx]: Remove commented code from pr

File: my-app/src/components/Pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  //       const arr = [];
+  //       for(let  = )
+  //     return Array(size).keys().map((num) => num + start)
+  //   }
+
+  //   if (totalPages <= pagesToShow) {
+
+  //   }
+
+  //   if (t)
```
Comment (TheForsakenSpirit) [inline my-app/src/components/Pagination.tsx]: remove

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

File: src/components/EvolutionsList.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        }
+      }
+    } catch (error) {
+      console.error(error);
+    }
+  }, [pokemonSpeciesURL]);
+
+  useEffect(() => {
+    fetchEvolutions();
+  }, [fetchEvolutions]);
```
Comment (TheForsakenSpirit) [inline src/components/EvolutionsList.tsx]: Just move this callback to `useEffect`

File: src/components/EvolutionsList.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  const fetchEvolutions = useCallback(async () => {
+    try {
+      const speciesResponse = await fetch(pokemonSpeciesURL);
+
+      if (speciesResponse.ok) {
+        const specie: unknown = await speciesResponse.json();
+
+        if (!isTypeOf<SpecieType>(specie, 'evolution_chain')) {
+          throw new Error('Received data is not Specie type');
+        }
```
Comment (TheForsakenSpirit) [inline src/components/EvolutionsList.tsx]: use not generic type guard.

File: src/components/PokemonCard.tsx [hunk 10]
```diff
@@ -1,17 +1,25 @@
...
   useEffect(() => {
     const fetchPokemonByid = async (id: number) => {
-      const response = await fetch(`${baseURL}/pokemon/${id}`);
-      console.log(response);
-      const data = (await response.json()) as PokemonType;
-      if ('weight' in data && 'base_experience' in data) {
+      try {
+        const response = await fetch(`${baseURL}/pokemon/${id}`);
+        console.log(response);
+        const data: unknown = (await response.json()) as PokemonType;
```
Comment (TheForsakenSpirit) [inline src/components/PokemonCard.tsx]: `unknown` or `PokemonType`?

File: src/components/PokemonCard.tsx [hunk 11]
```diff
@@ -1,17 +1,25 @@
...
+
+        if (!isTypeOf<PokemonType>(data, 'base_experience')) {
```
Comment (TheForsakenSpirit) [inline src/components/PokemonCard.tsx]: use not generic type guard

File: src/components/PokemonDetails.tsx [hunk 12]
```diff
@@ -8,9 +9,18 @@ export const PokemonDetails: FC<PokemonDetailsProps> = ({ id, baseURL }): JSX.El
...
   useEffect(() => {
     const fetchPokemonById = async (id: number) => {
-      const response = await fetch(`${baseURL}/pokemon/${id}`);
-      const data = (await response.json()) as PokemonType;
-      setPokemon(data);
+      try {
+        const response = await fetch(`${baseURL}/pokemon/${id.toString()}`);
+        const data: unknown = (await response.json()) as PokemonType;
+
+        if (!isTypeOf<PokemonType>(data, 'base_experience')) {
```
Comment (TheForsakenSpirit) [inline src/components/PokemonDetails.tsx]: use not generic type guard

File: src/utils/helpers.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+  if (typeof id === 'string') {
+    return parseInt(id);
+  }
+
+  return 0;
+};
+
+export const isTypeOf = <T extends Record<string, unknown>>(value: unknown, key: keyof T): value is T => {
+  return value instanceof Object && key in value;
```
Comment (TheForsakenSpirit) [inline src/utils/helpers.ts]: `key in value` potential dangerous condition.
If value = null it will throw an exception.

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

File: src/components/EvolutionsList.tsx (hunk 8 above)
Comment (apomelitos) [reply src/components/EvolutionsList.tsx]: Moved

File: src/components/EvolutionsList.tsx (hunk 9 above)
Comment (apomelitos) [reply src/components/EvolutionsList.tsx]: Replaced all type guards in ([ce3421c](https://github.com/apomelitos/typescript-test/pull/1/commits/ce3421c857d5f068257ec0ce7cfc7a6aa444006a)) commit

**Review by TheForsakenSpirit — APPROVED** [review]

**Pull request merged**